### PR TITLE
daemon: synchronously add endpoints into endpointmanager 

### PR DESCRIPTION
**Pull request by ianvernon**
Previously, \`regenerateRestoredEndpoints\` spawned a goroutine for each endpoint to be restored. In this goroutine, the endpoint to be restored was inserted into the endpointmanager. This meant that \`regenerateRestoredEndpoints\` could return before all goroutines were executed. As such, this meant that there was no guarantee of the state of endpointmanager after regenerateRestoredEndpoints returned. This could mean that functions which assume that endpoints are in the endpointmanager (e.g., `workloads.EnableEventListener`) would face a situation in which the endpoints were not always in the endpoint manager. By performing this insertion synchronously, we guarantee that all endpoints to be restored are inserted into the endpointmanager before `regenerateRestoredEndpoints` returns.

Signed-off by: Ian Vernon <ian@cilium.io>

Related-to : #4329 

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/4421)

### CODE REVIEW

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
Build failure: https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/3961/

VM provisioning failed during `go build` for k8s-1.10 tests.

**Comment by ianvernon** [comment]
test-me-please

**Review by joestringer — DISMISSED** [review]

File: daemon/state.go
Comment (manalibhutiyani) [inline daemon/state.go]: Quick question: Are the endpoints in the correct state (maybe `StateRestoring`) here when being inserted to `endpointmanager`. I see that you simply moved the insertion step out of the asynchronous go routine, but wanted to confirm that we do not need to do anything additional in terms of setting state to something like `StateRestoring` or equivalent.

Comment (ianvernon) [reply daemon/state.go]: `restored` is returned earlier by `d.restoreOldEndpoints`. `d.restoreOldEndpoints` invokes `readEPsFromDirNames`. `readEPsFromDirNames` calls `ParseEndpoint`, which sets `ep.state = StateRestoring`. Thus the endpoint is in `StateRestoring` when it is inserted into `endpointmanager` here. No further state changes are performed upon the endpoint or endpoints in `restored` after this point, until they are inserted into the `endpointmanager`; after they are inserted into the endpointmanager, there is no guarantee on the state of the endpoints, as some policy update event that is sent to the daemon could indeed force regeneration for endpoints, which in turn triggers a state change.

Comment (manalibhutiyani) [reply daemon/state.go]: thanks!

**Review by manalibhutiyani — DISMISSED** [review]

**Comment by ianvernon** [comment]
Build failure : https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/3962/

Needs investigation. Does not appear to be the issue this PR is attempting to fix. Testing again.

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
Another seemingly unrelated build failure:

```
/home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:373
CNI exec-plugins did not work correctly
Expected command: sudo -E PATH=$PATH:/opt/cni/bin -E CNI_PATH=/tmp/tmp.EfxZNopuWw/bin /tmp/tmp.EfxZNopuWw/cni/scripts/exec-plugins.sh add f30f3e4ac843f4fdc5e38d027a68687d2fe8e7e7df79aabec692becbe9b5e51b /proc/30225/ns/net 
To succeed, but it failed:
Exitcode: 1 
Stdout:
 	 crio-bridge : error executing ADD: 
	 
Stderr:
 	 Usage: bridge [ OPTIONS ] OBJECT { COMMAND | help }
	        bridge [ -force ] -batch filename
	 where	OBJECT := { link | fdb | mdb | vlan | monitor }
	 	OPTIONS := { -V[ersion] | -s[tatistics] | -d[etails] |
	 		     -o[neline] | -t[imestamp] | -n[etns] name |
	 		     -c[ompressvlans] -j{son} }
	 

/home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/runtime/connectivity.go:216
```

Link: https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/3963/testReport/junit/(root)/runtime/RuntimeValidatedConnectivityTest_With_CNI_Basic_connectivity_test/

I'm not sure how my changes could have caused this. This may be due to an update in the VM image used in the CI, which never actually passed tests: https://github.com/cilium/cilium/pull/4415

**Comment by ianvernon** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

**Comment by joestringer** [comment]
:lgtm:

---

Reviewed 1 of 2 files at r2.
Review status: all files reviewed, 1 unresolved discussion (waiting on @ianvernon)

---



*Comments from [Reviewable](https://reviewable.io/reviews/cilium/cilium/4421#-:-LEjuYHo1l_WWaGfxipc:bnfp4nl)*

**Review by aanm — APPROVED** [review]

**Review by manalibhutiyani — APPROVED** [review]

**Comment by ianvernon** [comment]
Merging; got approval from CODEOWNERS and the leftover stuff on reviewable has already been addressed.

**Pull request merged**
